modules/fundamental/src/vulnerability/service/mod.rs

```diff
@@ -155,6 +163,111 @@ impl VulnerabilityService {
 
         Ok(result.rows_affected)
     }
+
+    pub async fn analyze_purls<C>(
+        &self,
+        purls: Vec<&str>,
+        connection: &C,
+    ) -> Result<HashMap<String, Vec<VulnerabilityDetails>>, Error>
+    where
+        C: ConnectionTrait + StreamTrait,
+    {
+        if purls.is_empty() {
+            return Ok(Default::default());
+        }
+
+        let query = purls
+            .iter()
+            .filter_map(|p| {
+                let purl = Purl::from_str(p).map_err(Error::Purl);
+                let canonical = match purl {
+                    Ok(purl) => CanonicalPurl::from(purl),
```

Comment:
What is `CanonicalPurl` doing for you that `Purl` can't?

Reply:
Apparently nothing. I'll replace it. Thanks